test-utils/src/test_dir.rs

```diff
@@ -0,0 +1,134 @@
+    let mut rng = make_rng();
+    let mut delay = Duration::from_millis(1);
+    std::iter::from_fn(move || {
+        let item = rng.gen_range(Duration::ZERO..delay);
+        delay *= 2;
+        Some(item)
+    })
+    .take(10)
+}
+
+/// Create a new test root based on $CARGO_TARGET_TMPDIR/$CARGO_PKG_NAME/$custom_subdirs
+#[macro_export]
+macro_rules! test_root {
+    ($($dirs:expr),* $(,)?) => {
+        $crate::test_dir::TestRoot::create({
+            let mut path = ::std::path::PathBuf::from(env!("CARGO_TARGET_TMPDIR"));
+            path.push(env!("CARGO_PKG_NAME"));
+            $(path.push($dirs);)*
+            path
+        })
+    }
+}
+
+/// Root directory for a test run.
+///
+/// Corresponds to a working directory for a test binary run
+#[derive(Clone, Debug)]
+pub struct TestRoot(Arc<TestRootImpl>);
+
+impl TestRoot {
+    /// Create the root test directory
+    pub fn create(top_path: impl AsRef<Path>) -> io::Result<Self> {
+        // Create the top-level directory if it does not exist already
+        fs::create_dir_all(top_path.as_ref())?;
+
+        let mut rng = make_rng();
+
+        for delay in backoff_delays() {
+            let path = top_path.as_ref().join(format!("run_{:08x}", rng.gen::<u32>()));
+
+            // Attempt to create the candidate directory
+            match fs::create_dir(&path) {
+                Ok(()) => {
+                    // Sucesfully created
+                    let counter = AtomicU32::new(0);
+                    return Ok(Self(Arc::new(TestRootImpl { path, counter })));
+                }
+                Err(e) => match e.kind() {
+                    // If directory already exists, continue the loop to try again.
+                    std::io::ErrorKind::AlreadyExists => (),
+                    // Bail on any other error.
+                    _ => return Err(e),
+                },
+            }
+
+            std::thread::sleep(delay);
+        }
+
+        Err(io::Error::new(
+            io::ErrorKind::AlreadyExists,
+            "All attempted directory names exist",
+        ))
+    }
+
+    /// Create the new test subdirectory
+    pub fn fresh_test_dir(&self, name: impl AsRef<str>) -> TestDir {
+        let seq_no = self.0.counter.fetch_add(1, Ordering::SeqCst);
+        let path = self.0.path.join(format!("case_{:08x}_{}", seq_no, name.as_ref()));
+        fs::create_dir(&path).expect("directory creation to succeed");
+        TestDir { path }
+    }
+
+    /// Delete the directory. Panics if it is still in use.
+    pub fn delete(self) {
+        let inner = Arc::try_unwrap(self.0).expect("Test root still in use");
+        if let Err(err) = fs::remove_dir_all(&inner.path) {
+            eprintln!("Failed to remove test dir {:?}: {}", &inner.path, err);
+        }
+    }
+}
+
+#[derive(Debug)]
+struct TestRootImpl {
+    path: PathBuf,
+    counter: AtomicU32,
+}
+
+/// Represents a test directory. It gets deleted if the object is dropped, unless there's a panic.
```

Comment:
I think this comment isn't quite accurate. We don't delete individual directories, we delete everything that is located inside the root directory. Additionally the deletion isn't performed when this object is dropped. Am I missing something?

Reply:
Yeah, I attempted what's written there but it proved rather difficult to do with how the current storage tests work. Will update the comment. Thanks for pointing this out